saleor/order/fetch.py

```diff
@@ -147,4 +153,61 @@ def fetch_draft_order_lines_info(
     ):
         voucher_info = fetch_voucher_info(voucher, order.voucher_code)
         apply_voucher_to_line(voucher_info, lines_info)
+        denormalized_voucher_info = _get_denormalized_voucher_info(lines_info, voucher)
+        _apply_denormalized_voucher_to_line_info(
+            lines_info, denormalized_voucher_info, order.voucher_code
+        )
     return lines_info
+
+
+def _get_denormalized_voucher_info(lines_info: list[EditableOrderLineInfo], voucher):
+    voucher_discounts = [
+        discount
+        for line_info in lines_info
+        for discount in line_info.discounts
+        if discount.voucher == voucher
+    ]
+    if not voucher_discounts:
+        return None
+
+    voucher_discount = voucher_discounts[0]
+    return VoucherDenormalizedInfo(
+        discount_value=voucher_discount.value,
+        discount_value_type=voucher_discount.value_type,
+        voucher_type=voucher.type,
+        reason=voucher_discount.reason,
+        name=voucher_discount.name,
+        apply_once_per_order=voucher.apply_once_per_order,
+        origin_line_id=voucher_discount.line_id,
+    )
+
+
+def _apply_denormalized_voucher_to_line_info(
+    lines_info: list[EditableOrderLineInfo],
+    denormalized_voucher_info: VoucherDenormalizedInfo | None,
+    voucher_code: str | None,
+):
+    if not denormalized_voucher_info:
+        return
+
+    # Denormalized vouchers of type SPECIFIC PRODUCT shouldn't be evaluated against
+    # the latest eligible product catalog so it should be applied to the same line
+    # it originates from
+    if denormalized_voucher_info.voucher_type == VoucherType.SPECIFIC_PRODUCT:
+        if line_info := next(
+            line_info
+            for line_info in lines_info
+            if line_info.line.pk == denormalized_voucher_info.origin_line_id
+        ):
+            line_info.voucher_denormalized_info = denormalized_voucher_info
+            line_info.voucher_code = voucher_code
+        return
+
+    # Denormalized voucher applicable once per order can be applied to the different
+    # line than it originates from (it depends on the actual cheapest line)
+    if denormalized_voucher_info.apply_once_per_order is True:
```

Comment:
Isn't like it should be `elif`? Right now you will call the first `if`, then the second one.

Reply:
Right, missed that